module/zfs/arc.c

```diff
@@ -5576,6 +5610,20 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 				    arc_buf_hdr_t *, hdr);
 				ARCSTAT_BUMP(arcstat_async_upgrade_sync);
 			}
+			/*
+			 * Cache-only lookups should only occur from consumers
+			 * that do not have any data yet.  However, prefetch
+			 * I/O of this block could be in progress.  Since
+			 * cache-only lookups must be synchronous, the done
+			 * callback chaining cannot occur here.  In that case,
+			 * simply return as a cache miss.
+			 */
+			if (cached_only) {
+				*arc_flags &= ~ARC_FLAG_CACHED;
+				mutex_exit(hash_lock);
+				done(NULL, zb, bp, NULL, private);
+				return (0);
```

Comment:
Let's set `rc` and `goto out;` like the rest of the function.  I'm also confused why we're returning 0 instead of say `ENOENT` to indicate we failed to find it in the cache (since it was still in flight).